examples/pytorch/resnet.py

```diff
@@ -0,0 +1,187 @@
+"""This is a simple example of how to use Mountpoint as a PyTorch data loader via the torchdata library.
+It trains a ResNet-50 model for a few epochs using PyTorch Lightning, with synthetic ImageNet-sized
+training data, stored in S3 as shards in WebDataset format.
+
+Run it like this to upload the training shards to an S3 bucket:
+
+    python resnet.py make s3://DOC-EXAMPLE-BUCKET/shard-data/ --num-images 10000
+
+And then run it like this to run the training loop:
+
+    python resnet.py train s3://DOC-EXAMPLE-BUCKET/shard-data/ --source-kind mountpoint --batch-size 256
+
+The --source-kind argument controls how the data is loaded from S3:
+* `mountpoint` spawns a Mountpoint instance and accesses it as a local file system
+* `s3io` uses the S3FileLoader datapipe from torchdata (aka the PyTorch plugin for Amazon S3)
+* `fsspec` uses the FSSpecFileOpener datapipe from torchdata (which uses the fsspec and s3fs libraries)
+"""
+
+import argparse
+import atexit
+import boto3
+import lightning as L
+import os
+from PIL import Image
+import subprocess
+import tempfile
+import time
+import torch
+import torchdata
+import torchvision
+from typing import *
+import webdataset as wds
+
+
+def make_sharded_dataset(bucket: str, prefix: str, num_images: int, max_shard_size: int):
+    """Make a fake dataset in WebDataset format and upload it to S3. In reality you'd already have
+    this dataset in S3, so this is just for benchmarking purposes. We choose the image sizes to be
+    roughly ImageNet-esque."""
+
+    s3 = boto3.client("s3")
+    ds = torchvision.datasets.FakeData(size=num_images, image_size=(3, 224, 224), num_classes=100)
+
+    with tempfile.TemporaryDirectory() as tempdir:
```

Comment:
Would be nice to have the with / without shared benchmark. This requires code changes and this isn't always an assumption to be made about user dataset.

Reply:
Agreed! This is just a first example, but we're looking at non-sharded versions too.